core/include/drivers/stm32_gpio.h

```diff
@@ -0,0 +1,163 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+ * Copyright (c) 2017-2019, STMicroelectronics
+ *
+ * STM32 GPIO driver relies on platform util fiunctions to get base address
+ * and clock ID of the GPIO banks. The drvier API allows to retrieve pin muxing
+ * configuration for given nodes and load them at runtime. A pin control
+ * instance provide an active and a standby configuration. Pin onwer is
+ * responsible to load to expected configuration during PM state transitions
+ * as STM32 GPIO driver does no register callbacks to the PM framework.
+ */
+
+#ifndef __STM32_GPIO_H
+#define __STM32_GPIO_H
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <stddef.h>
+
+#define GPIO_MODE_INPUT		0x0
+#define GPIO_MODE_OUTPUT	0x1
+#define GPIO_MODE_ALTERNATE	0x2
+#define GPIO_MODE_ANALOG	0x3
+
+#define GPIO_OTYPE_PUSH_PULL	0x0
+#define GPIO_OTYPE_OPEN_DRAIN	0x1
+
+#define GPIO_OSPEED_LOW		0x0
+#define GPIO_OSPEED_MEDIUM	0x1
+#define GPIO_OSPEED_HIGH	0x2
+#define GPIO_OSPEED_VERY_HIGH	0x3
+
+#define GPIO_PUPD_NO_PULL	0x0
+#define GPIO_PUPD_PULL_UP	0x1
+#define GPIO_PUPD_PULL_DOWN	0x2
+
+#define GPIO_OD_LEVEL_LOW	0x0
+#define GPIO_OD_LEVEL_HIGH	0x1
+
+/*
+ * SoC configuration for a GPIO pin, fits in a 16bit word
+ * @mode: One of GPIO_MODE_*
+ * @otype: One of GPIO_OTYPE_*
+ * @ospeed: One of GPIO_OSPEED_*
+ * @pupd: One of GPIO_PUPD_*
+ * @od: One of GPIO_OD_*
+ * @af: Alternate function numerical ID between 0 and 15
+ */
+
+struct gpio_cfg {
```

Comment:
Does this depend on memory layout of the struct?
If so then this isn't entirely portable since the C standard doesn't define the order in which the bits are assigned.

Reply:
No, this is local to this driver. This structure is used internally by the driver to save/restore GPIO/pinmux configurations to/from RAM. This structure definition is there only to make a single GPIO/pinmux config to fit in a 16bit RAM cell.

Reply:
Good, a short comment stating that would be welcome.